### ISSUE TITLE
[JENKINS-3966] Add option to clean workspace before each build

### ISSUE THREAD

**Issue by timja** [timja/jenkins-gh-issues-poc-06-20#49]
<p>If this option is selected the workspace will be deleted before each build<br/>
begins so that each build can have a fresh workspace.  This will be off by default.</p>

---
<details><summary><i>Originally reported by <a title="kaxelson" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kaxelson">kaxelson</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-3966" target="_blank">Add option to clean workspace before each build</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="kaxelson" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kaxelson">kaxelson</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2014-05-23T15:20:46+00:00
<li><b>imported</b>: 2022/01/10
</ul></i>
</details>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kaxelson">kaxelson</a>:</i>
<p>starting work on this</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in hudson<br/>
User: : kaxelson<br/>
Path:<br/>
 trunk/hudson/main/core/src/main/java/hudson/model/AbstractBuild.java<br/>
 trunk/hudson/main/core/src/main/java/hudson/model/AbstractProject.java<br/>
 trunk/hudson/main/core/src/main/resources/hudson/model/AbstractItem/configure-common.jelly<br/>
 trunk/hudson/main/core/src/main/resources/lib/hudson/project/config-cleanWorkspace.jelly<br/>
 trunk/hudson/main/war/resources/help/project-config/cleanWorkspace.html<br/>
<a href="http://fisheye4.cenqua.com/changelog/hudson/?cs=19373" class="external-link" target="_blank" rel="nofollow noopener">http://fisheye4.cenqua.com/changelog/hudson/?cs=19373</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-3966&#93;</span> added option to clean workspace before each build as part of advanced project configuration</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in hudson<br/>
User: : kaxelson<br/>
Path:<br/>
 trunk/www/changelog.html<br/>
<a href="http://fisheye4.cenqua.com/changelog/hudson/?cs=19374" class="external-link" target="_blank" rel="nofollow noopener">http://fisheye4.cenqua.com/changelog/hudson/?cs=19374</a><br/>
Log:<br/>
<a href="https://issues.jenkins.io/browse/JENKINS-3966" title="Add option to clean workspace before each build" class="issue-link" data-issue-key="JENKINS-3966"><del>JENKINS-3966</del></a>: adding changelog message</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Rolled back from 1.315 release in rev.19544. See<br/>
<a href="http://www.nabble.com/attention-all-subversion-users-td24335693.html" class="external-link" target="_blank" rel="nofollow noopener">http://www.nabble.com/attention-all-subversion-users-td24335693.html</a> for the<br/>
discussion.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Note that the Mercurial plugin already has an option to delete all nonversioned<br/>
files from the workspace before starting. (The .hg/ subdirectory is excluded as<br/>
is usual.) I think there is a similar issue open for Subversion. This is more<br/>
efficient than deleting everything in the workspace, but may not be applicable<br/>
for all situations.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=dma_k">dma_k</a>:</i>
<p>The option to remove all workspace is important to make maven release plugin to<br/>
work correctly for the projects, that generate sources e.g. with JAXB. In my<br/>
case I commit JAXB sources from Windows environment, which after being<br/>
re-generated have different EOF and SVN senses this as uncommitted change, which<br/>
in its turn blocks the release.<br/>
If the fix is ready, what is the target milestone for this fix?</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=smr99">smr99</a>:</i>
<p>What is the status of this?  I read through the referenced email thread (now at <a href="http://markmail.org/message/mcgftdni6cvpca7v" class="external-link" target="_blank" rel="nofollow noopener">http://markmail.org/message/mcgftdni6cvpca7v</a>) which seemed to end with agreement on the feature but issues with backwards compatibility.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jhoomsharabi">jhoomsharabi</a>:</i>
<p>Does this functionality anytime implemented, as I see this issue is being reopened. </p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rpetti">rpetti</a>:</i>
<p>Can't this option just call the same function as the "Wipe Out Workspace" feature? If the SCM plugin supports it, then everything will be fine. If it doesn't then it simply won't delete the workspace.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=bvakili">bvakili</a>:</i>
<p>I like Kohsuke's proposal of providing an 'extension point in core to abstract away various "clean" logic'.<br/>
<a href="http://n4.nabble.com/attention-all-subversion-users-tp393646p393656.html" class="external-link" target="_blank" rel="nofollow noopener">http://n4.nabble.com/attention-all-subversion-users-tp393646p393656.html</a></p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 core/src/main/java/hudson/scm/NoCleaner.java<br/>
 core/src/main/java/hudson/scm/RmRfCleaner.java<br/>
 core/src/main/java/hudson/scm/WorkspaceCleaner.java<br/>
 core/src/main/java/hudson/scm/WorkspaceCleanerDescriptor.java<br/>
 core/src/main/resources/hudson/scm/SCM/configure-workspaceCleaner.jellytag<br/>
 core/src/main/resources/hudson/scm/WorkspaceCleaner/config.jelly<br/>
<a href="http://jenkins-ci.org/commit/jenkins/07c6c4f1dc9c41a2cf8fc1ee9a8935310ca9afb0" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/07c6c4f1dc9c41a2cf8fc1ee9a8935310ca9afb0</a><br/>
Log:<br/>
<a href="https://issues.jenkins.io/browse/JENKINS-3966" title="Add option to clean workspace before each build" class="issue-link" data-issue-key="JENKINS-3966"><del>JENKINS-3966</del></a> added a new extension point for workspace cleaner</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jlpinardon">jlpinardon</a>:</i>
<p>It would be a user friendly enhancement to add the ability to select a list (possibly based on pattern) of directory to  wipe out (or not).</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=johnstosh">johnstosh</a>:</i>
<p>(I've been wiping out specific directories by adding ant tasks into the my jobs.)</p>

<p>Hey guys, thanks for enhancing this section of Jenkins.  The clean selections in Jenkins 1.414 work great for me: both the delete-workspace and the other varieties of cleaning.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ganncamp">ganncamp</a>:</i>
<p>I think this issue has been addressed by this new-ish plugin<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/help_16.png" height="16" width="16" align="absmiddle" alt="" border="0"/> <a href="https://wiki.jenkins-ci.org/display/JENKINS/Workspace+Cleanup+Plugin" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Workspace+Cleanup+Plugin</a></p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Workspace Cleanup Plugin would delete the entire workspace. For most purposes you would rather just delete unversioned files (typically build products), as well as reverting any local modifications to versioned files, since a full checkout/clone/etc. is likely to be a lot more expensive than an incremental update.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>The mentioned plugin should be used; or many SCM plugins now offer the option of deleting unversioned files.</p>

**Comment by timja** [comment]
<i>[Originally related to: <a href="https://github.com/timja/jenkins-gh-issues-poc-06-20/issues?q=in%3Atitle%20JENKINS-3934">JENKINS-3934</a>]</i>
